### PR TITLE
feat(synthetics): validated staged assets

### LINKED ISSUE #11630: [synthetics] Validation of canary file existence done too early
When creating Canary lambda asset resource, the bundling does not happen before the validation of "existence" of canary file is done. This blocks the usage of bundling options in many case.

### Reproduction Steps

```
    new Canary(this, props.canaryId, {
      canaryName: 'testcanary',
      test: Test.custom({
        code: Code.fromAsset(path.join(__dirname, '../canaryScripts/), {
          bundling: {
            image: lambda.Runtime.NODEJS_12_X.bundlingDockerImage,
            command: [
              'bash',
              '-c',
              [
                `mkdir -p /asset-output/nodejs/node_modules/`,
                `cp -R /asset-input/* /asset-output/nodejs/node_modules/`,
              ].join(' && '),
            ],
            user: 'root',
          },
        }),
        handler: `canary.handler`,
      }),
      runtime: Runtime.SYNTHETICS_NODEJS_2_0,
    });
```

### What did you expect to happen?

I expected that the bundling would happen before CDK validates if canary file has been found from correct location of zip-file generated.

Using bundling I can run optimisation, validations, include only needed files and re-structure my code asset before creating zip file.

### What actually happened?

I get error message before bundling has been executed as following: `The canary resource requires that the handler is present at "nodejs/node_modules/canary.js" but not found at /Users/mypath/src/synthetics/ (https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries_WritingCanary.html#CloudWatch_Synthetics_Canaries_write_from_scratch)
Subprocess exited with error 1`

### Environment

  - **CDK CLI Version  : 1.74.0**
  - **Framework Version: 1.74.0**
  - **Node.js Version: v12.18.4** 
  - **OS               :  macOS 11.0.1**
  - **Language (Version): Typescript** 

### Other

The code that triggers error is here:  https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-synthetics/lib/code.ts#L135

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by RichiCoder1**
This changes synthetic to validated staged assets instead of raw assets. This most importantly allows for asset bundling (if provided) to kick in before validating assets.

Wasn't sure whether there should be a readme change, since this is just making the construct play nice with `Asset`'s bundling feature.

Fixes #11630 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by RichiCoder1** [comment]
It seems a stretch for this L2 to get a `lambda-nodejs` equivalent, so wanted to do this to at least open the door to community constructs to fill this gap. Currently you have to do bundling outside synth as a pre-build step.

File: packages/@aws-cdk/aws-synthetics/lib/code.ts [hunk 1]
```diff
@@ -107,6 +106,8 @@ export class AssetCode extends Code {
       });
     }
 
+    this.validateCanaryAsset(scope, handler, family);
```
Comment (RichiCoder1) [inline packages/@aws-cdk/aws-synthetics/lib/code.ts:109]: I'm not sure if there was a specific reason for this ordering, but looks like this logic predates Asset's bundling support so there doesn't appear to be an issue doing this now.

File: packages/@aws-cdk/aws-synthetics/lib/code.ts [hunk 2]
```diff
@@ -127,18 +128,25 @@ export class AssetCode extends Code {
...
+    // Get the staged (or copied) asset path.
+    const assetOutdir = cdk.Stage.of(scope)?.assetOutdir;
+    let assetPath = this.assetPath;
+    if (assetOutdir) {
+      assetPath = path.join(assetOutdir, this.asset!.assetPath);
+    }
+
     if (path.extname(this.assetPath) !== '.zip') {
-      if (!fs.lstatSync(this.assetPath).isDirectory()) {
+      if (this.asset!.isFile) {
```
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-synthetics/lib/code.ts:140]: I understand that with the current ordering of things, `this.asset` exists, but this is a detail that we should abstract out of the function. I prefer if we never use `!` and instead just make sure that `this.asset` exists

```ts
if (this.asset && this.asset.isFile) {}
```

File: packages/@aws-cdk/aws-synthetics/lib/code.ts [hunk 3]
```diff
@@ -127,18 +128,25 @@ export class AssetCode extends Code {
...
    * @param handler the canary handler
    */
-  private validateCanaryAsset(handler: string, family: RuntimeFamily) {
+  private validateCanaryAsset(scope: Construct, handler: string, family: RuntimeFamily) {
+    // Get the staged (or copied) asset path.
+    const assetOutdir = cdk.Stage.of(scope)?.assetOutdir;
+    let assetPath = this.assetPath;
+    if (assetOutdir) {
+      assetPath = path.join(assetOutdir, this.asset!.assetPath);
+    }
```
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-synthetics/lib/code.ts:137]:
```suggestion
    const assetPath = assetOutdir ? path.join(assetOutdir, this.asset.assetPath) : this.assetPath;
```

Comment (kaizen3031593) [inline packages/@aws-cdk/aws-synthetics/lib/code.ts:137]:
```suggestion
    const assetPath = assetOutdir ? path.join(assetOutdir, this.asset.assetPath) : this.assetPath;
```

File: packages/@aws-cdk/aws-synthetics/lib/code.ts [hunk 4]
```diff
@@ -127,18 +128,25 @@ export class AssetCode extends Code {
    *
    * @param handler the canary handler
    */
-  private validateCanaryAsset(handler: string, family: RuntimeFamily) {
+  private validateCanaryAsset(scope: Construct, handler: string, family: RuntimeFamily) {
```
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-synthetics/lib/code.ts:131]: I suggest you add this at the beginning of this function:

```ts
if (!this.asset) {
    throw new Error("'validateCanaryAsset' must be called after 'this.asset' is defined");
}

Comment (kaizen3031593) [inline packages/@aws-cdk/aws-synthetics/lib/code.ts:131]: I suggest you add this at the beginning of this function:

```ts
if (!this.asset) {
    throw new Error("'validateCanaryAsset' must be called after 'this.asset' is defined");
}

File: packages/@aws-cdk/aws-synthetics/lib/code.ts (hunk 1 above)
Comment (kaizen3031593) [reply packages/@aws-cdk/aws-synthetics/lib/code.ts:109]: I think the ordering just has to do with the principle of throwing errors as early as possible.

**Review by kaizen3031593 — CHANGES_REQUESTED** [review]
Hi @RichiCoder1, some preliminary thoughts on this one. Looks good so far.

File: packages/@aws-cdk/aws-synthetics/lib/code.ts (hunk 4 above)
Comment (RichiCoder1) [reply packages/@aws-cdk/aws-synthetics/lib/code.ts:131]: Seems reasonable and a better gaurd!

**Comment by RichiCoder1** [comment]
@kaizencc howdy! It's been a while, just wanted to check in on this :)

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
Apologies for the long wait on this one. Unfortunately, our move over to v2 caused this to get quite messy with conflicts. If you are still interested in working on this and have the capacity to resolve them, we'll take another look at this.

**Comment by RichiCoder1** [comment]
Absolutely! I'll take a swing this weekend and getting this rebased

**Pull request closed without merging**